File: python/A_star.py
import heapq
from typing import List, Tuple, Optional

def heuristic(a: Tuple[int, int], b: Tuple[int, int]) -> int:
    """Manhattan distance heuristic"""
    return abs(a[0] - b[0]) + abs(a[1] - b[1])

def astar(grid: List[List[int]], 
          start: Tuple[int, int], 
          goal: Tuple[int, int]) -> Optional[List[Tuple[int, int]]]:
    """
    A* pathfinding algorithm
    
    Args:
        grid: 2D grid where 0 is walkable, 1 is wall
        start: Starting position (row, col)
        goal: Goal position (row, col)
    
    Returns:
        List of positions forming the path, or None if no path exists
    """
    rows, cols = len(grid), len(grid[0])
    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    
    # Priority queue: (f_score, g_score, position)
    open_set = [(heuristic(start, goal), 0, start)]
    came_from = {}
    g_score = {start: 0}
    
    while open_set:
        _, curr_g, curr = heapq.heappop(open_set)
        
        if curr == goal:
            # Reconstruct path
            path = [goal]
            while path[-1] in came_from:
                path.append(came_from[path[-1]])
            return path[::-1]
        
        # Skip if we've found a better path already
        if curr_g > g_score.get(curr, float('inf')):
            continue
        
        for dx, dy in directions:
            neighbor = (curr[0] + dx, curr[1] + dy)
            
            # Check bounds and walls
            if (0 <= neighbor[0] < rows and 
                0 <= neighbor[1] < cols and 
                grid[neighbor[0]][neighbor[1]] == 0):
                
                tentative_g = curr_g + 1
                
                if tentative_g < g_score.get(neighbor, float('inf')):
                    came_from[neighbor] = curr
                    g_score[neighbor] = tentative_g
                    f_score = tentative_g + heuristic(neighbor, goal)
                    heapq.heappush(open_set, (f_score, tentative_g, neighbor))
    
    return None

def print_grid_with_path(grid: List[List[int]], path: List[Tuple[int, int]]):
    """Visualize the grid with the path"""
    path_set = set(path)
    for i, row in enumerate(grid):
        for j, cell in enumerate(row):
            if (i, j) in path_set:
                if (i, j) == path[0]:
                    print("S", end=" ")
                elif (i, j) == path[-1]:
                    print("G", end=" ")
                else:
                    print("*", end=" ")
            elif cell == 1:
                print("█", end=" ")
            else:
                print("·", end=" ")
        print()

if __name__ == "__main__":
    grid = [
        [0, 0, 0, 0, 0],
        [0, 1, 1, 1, 0],
        [0, 0, 0, 0, 0],
        [0, 1, 0, 1, 0],
        [0, 0, 0, 0, 0],
    ]
    
    start = (0, 0)
    goal = (4, 4)
    
    path = astar(grid, start, goal)
    
    if path:
        print("Path found:")
        for pos in path:
            print(f"({pos[0]},{pos[1]})", end=" ")
        print(f"\nSteps: {len(path) - 1}\n")
        
        print("Grid visualization:")
        print_grid_with_path(grid, path)
    else:
        print("No path found")

File: go/A_star.go
package main

import (
	"container/heap"
	"fmt"
	"math"
)

type Point struct {
	x, y int
}

type Node struct {
	pos    Point
	g, h   float64
	parent *Node
	idx    int
}

func (n *Node) f() float64 { return n.g + n.h }

type PriorityQueue []*Node

func (pq PriorityQueue) Len() int           { return len(pq) }
func (pq PriorityQueue) Less(i, j int) bool { return pq[i].f() < pq[j].f() }
func (pq PriorityQueue) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
	pq[i].idx, pq[j].idx = i, j
}

func (pq *PriorityQueue) Push(x interface{}) {
	n := len(*pq)
	node := x.(*Node)
	node.idx = n
	*pq = append(*pq, node)
}

func (pq *PriorityQueue) Pop() interface{} {
	old := *pq
	n := len(old)
	node := old[n-1]
	old[n-1] = nil
	node.idx = -1
	*pq = old[0 : n-1]
	return node
}

func heuristic(a, b Point) float64 {
	return math.Abs(float64(a.x-b.x)) + math.Abs(float64(a.y-b.y))
}

func AStar(grid [][]int, start, goal Point) []Point {
	rows, cols := len(grid), len(grid[0])
	dirs := []Point{{0, 1}, {1, 0}, {0, -1}, {-1, 0}}

	pq := &PriorityQueue{}
	heap.Init(pq)

	startNode := &Node{pos: start, g: 0, h: heuristic(start, goal)}
	heap.Push(pq, startNode)

	visited := make(map[Point]bool)
	costs := make(map[Point]float64)
	costs[start] = 0

	for pq.Len() > 0 {
		curr := heap.Pop(pq).(*Node)

		if curr.pos == goal {
			path := []Point{}
			for n := curr; n != nil; n = n.parent {
				path = append([]Point{n.pos}, path...)
			}
			return path
		}

		if visited[curr.pos] {
			continue
		}
		visited[curr.pos] = true

		for _, d := range dirs {
			nx, ny := curr.pos.x+d.x, curr.pos.y+d.y
			neighbor := Point{nx, ny}

			if nx < 0 || ny < 0 || nx >= rows || ny >= cols || grid[nx][ny] == 1 {
				continue
			}

			newG := curr.g + 1

			if cost, ok := costs[neighbor]; !ok || newG < cost {
				costs[neighbor] = newG
				node := &Node{
					pos:    neighbor,
					g:      newG,
					h:      heuristic(neighbor, goal),
					parent: curr,
				}
				heap.Push(pq, node)
			}
		}
	}

	return nil
}

func main() {
	grid := [][]int{
		{0, 0, 0, 0, 0},
		{0, 1, 1, 1, 0},
		{0, 0, 0, 0, 0},
		{0, 1, 0, 1, 0},
		{0, 0, 0, 0, 0},
	}

	start := Point{0, 0}
	goal := Point{4, 4}

	path := AStar(grid, start, goal)

	if path != nil {
		fmt.Println("Path found:")
		for _, p := range path {
			fmt.Printf("(%d,%d) ", p.x, p.y)
		}
		fmt.Printf("\nSteps: %d\n", len(path)-1)
	} else {
		fmt.Println("No path found")
	}
}
